### PR TITLE
fs/operations: collect operation objects concurrently

### PR DESCRIPTION
**Pull request by zhucan**
Signed-off-by: zhuc <zhucan.k8s@gmail.com>

#### What is the purpose of this change?
support multi-threads to head dst object with flag '--files-from' and '--no-traverse'

#### Was the change discussed in an issue or in the forum before?
#4808 

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request).
- [ ] I have added tests for all changes in this PR if appropriate.
- [ ] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

**Comment by zhucan** [comment]
Failed to compress rclone?

**Comment by darthShadow** [comment]
The error is pretty clear:

```
# github.com/rclone/rclone/fs/march
fs/march/march.go:420:29: undefined: fs.Config
```

This syntax was changed in the master branch recently: https://github.com/rclone/rclone/commit/2e21c58e6a6c69be8c2c46cda157d977388687c5

Please update your usage as shown in the commit which should get your builds compiling again.

**Comment by zhucan** [comment]
> The error is pretty clear:
> 
> ```
> # github.com/rclone/rclone/fs/march
> fs/march/march.go:420:29: undefined: fs.Config
> ```
> 
> This syntax was changed in the master branch recently: [2e21c58](https://github.com/rclone/rclone/commit/2e21c58e6a6c69be8c2c46cda157d977388687c5)
> 
> Please update your usage as shown in the commit which should get your builds compiling again.

Thanks, I'm not update the master branch.

**Comment by ivandeex** [comment]
It's generally advisable to try at least `go build` and tests locally before pushing. See https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request

**Comment by ivandeex** [comment]
For note: a.t.m. we have pending https://github.com/rclone/rclone/pull/4821 that seeks to adds more concurrency using `--checkers` and https://github.com/rclone/rclone/pull/4172 that limits concurrency stronger using `--s3-concurrency`.

File: fs/march/march.go [hunk 1]
```diff
@@ -416,17 +417,29 @@ func (m *March) processJob(job listDirJob) ([]listDirJob, error) {
 	}
 
 	// If NoTraverse is set, then try to find a matching object
-	// for each item in the srcList
+	// for each item in the srcList with threads
```
Comment (ivandeex) [inline fs/march/march.go:420]: `srcList with threads` sounds strange

**Comment by zhucan** [comment]
> e have pending #482

Is it needed to add other flags ?

File: fs/march/march.go (hunk 1 above)
Comment (zhucan) [reply fs/march/march.go:420]: sounds good.

Comment (zhucan) [reply fs/march/march.go:420]: like this 'for each item in the srcList to head dst object' is ok?

**Comment by ivandeex** [comment]
What rclone comnands are affected by `processJobs`?
https://github.com/rclone/rclone/blob/979bb07c86f9680466e9421da116aaf3af94c1da/fs/march/march.go#L373

**Comment by zhucan** [comment]
> What rclone commands are affected by `processJob` that is modified by this PR?
> https://github.com/rclone/rclone/blob/979bb07c86f9680466e9421da116aaf3af94c1da/fs/march/march.go#L373

rclone copy, move, sync (with flags '--files-from')and so on

File: fs/march/march.go [hunk 2]
```diff
@@ -416,17 +417,29 @@ func (m *March) processJob(job listDirJob) ([]listDirJob, error) {
 	}
 
 	// If NoTraverse is set, then try to find a matching object
-	// for each item in the srcList
+	// for each item in the srcList to head dst object
+	ci := fs.GetConfig(m.Ctx)
+	limiter := make(chan bool, ci.Checkers)
```
Comment (ivandeex) [inline fs/march/march.go:422]: `bool` is 4 data bytes, `struct{}` is 0 data bytes
let's save a little ram

File: fs/march/march.go [hunk 3]
```diff
@@ -416,17 +417,29 @@ func (m *March) processJob(job listDirJob) ([]listDirJob, error) {
...
 	if m.NoTraverse && !m.NoCheckDest {
 		for _, src := range srcList {
-			if srcObj, ok := src.(fs.Object); ok {
-				leaf := path.Base(srcObj.Remote())
-				dstObj, err := m.Fdst.NewObject(m.Ctx, path.Join(job.dstRemote, leaf))
-				if err == nil {
-					dstList = append(dstList, dstObj)
+			wg.Add(1)
+			limiter <- true
```
Comment (ivandeex) [inline fs/march/march.go:426]: please use `struct{}{}` instead of `true`

File: fs/march/march.go (hunk 2 above)
Comment (zhucan) [reply fs/march/march.go]: done

File: fs/march/march.go (hunk 3 above)
Comment (zhucan) [reply fs/march/march.go]: done

File: fs/march/march.go [hunk 4]
```diff
@@ -416,17 +417,29 @@ func (m *March) processJob(job listDirJob) ([]listDirJob, error) {
...
+						mutex.Lock()
+						dstList = append(dstList, dstObj)
+						mutex.Unlock()
+					}
 				}
-			}
+				<-limiter
+			}(limiter, src)
 		}
+		close(limiter)
```
Comment (ncw) [inline fs/march/march.go:441]: You don't want to close the limiter - just remove this line.

**Review by ncw — COMMENTED** [review]
Thank you this is looking very good :-)

A couple of minor things inline.

Thank you

File: fs/march/march.go [hunk 5]
```diff
@@ -416,17 +417,29 @@ func (m *March) processJob(job listDirJob) ([]listDirJob, error) {
 	}
 
 	// If NoTraverse is set, then try to find a matching object
-	// for each item in the srcList
+	// for each item in the srcList with threads
+	ci := fs.GetConfig(m.Ctx)
+	limiter := make(chan bool, ci.Checkers)
```
[the remark this answers is not in the document]
Comment (ncw) [reply fs/march/march.go]: I think you've done this already

File: fs/march/march.go [hunk 6]
```diff
@@ -416,17 +417,29 @@ func (m *March) processJob(job listDirJob) ([]listDirJob, error) {
...
 	if m.NoTraverse && !m.NoCheckDest {
 		for _, src := range srcList {
-			if srcObj, ok := src.(fs.Object); ok {
-				leaf := path.Base(srcObj.Remote())
-				dstObj, err := m.Fdst.NewObject(m.Ctx, path.Join(job.dstRemote, leaf))
-				if err == nil {
-					dstList = append(dstList, dstObj)
+			wg.Add(1)
+			limiter <- true
```
[the remark this answers is not in the document]
Comment (ncw) [reply fs/march/march.go]: and this

File: fs/march/march.go [hunk 7]
```diff
@@ -382,6 +382,7 @@ func (m *March) processJob(job listDirJob) ([]listDirJob, error) {
 		srcList, dstList       fs.DirEntries
 		srcListErr, dstListErr error
 		wg                     sync.WaitGroup
+		mutex                  sync.Mutex
```
[the remark this answers is not in the document]
Comment (zhucan) [reply fs/march/march.go]: done

File: fs/march/march.go (hunk 4 above)
Comment (zhucan) [reply fs/march/march.go]: done

**Comment by ncw** [comment]
Perfect - thank you :-)

I'll merge this how.

**Pull request merged**